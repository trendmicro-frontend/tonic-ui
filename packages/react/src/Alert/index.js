import React, { forwardRef } from 'react';
import Box from '../Box';
import ButtonBase from '../ButtonBase';
import Flex from '../Flex';
import Icon from '../Icon';
import Space from '../Space';
import {
  useAlertStyle,
  useAlertIconStyle,
  useAlertMessageStyle,
  useAlertCloseButtonStyle,
} from './styles';

const defaultVariant = 'solid';
const defaultSeverity = 'success';

const getIconBySeverity = (severity) => {
  const iconName = {
    success: 'success',
    info: 'info',
    warning: 'warning-triangle',
    error: 'error',
  }[severity];

  if (!iconName) {
    return null;
  }

  return (
    <Icon icon={`${iconName}`} />
  );
};

const AlertIcon = (props) => (
  <Flex {...props} />
);

const AlertMessage = (props) => (
  <Box {...props} />
);

const AlertCloseButton = (props) => (
  <ButtonBase {...props} />
);

const Alert = forwardRef((
  {
    isClosable: _isClosable = false,
    isCloseButtonVisible: LEGACY_isCloseButtonVisible = false, // eslint-disable-line camelcase
    onClose,
    variant = defaultVariant,
    severity = defaultSeverity,
    icon,
    children,
    ...rest
  },
  ref,
) => {
  const isClosable = _isClosable || LEGACY_isCloseButtonVisible; // eslint-disable-line camelcase
  const styleProps = useAlertStyle({ variant, severity });
  const iconStyleProps = useAlertIconStyle({ variant, severity });
  const messageStyleProps = useAlertMessageStyle();
  const closeButtonStyleProps = useAlertCloseButtonStyle({ variant });

  if (typeof icon === 'string') {
    icon = (<Icon icon={icon} />);
  }
  if (typeof icon === 'undefined') {
    icon = getIconBySeverity(severity);
  }

  return (
    <Flex
      ref={ref}
      align="flex-start"
      justify="space-between"
      {...styleProps}
      {...rest}
    >
      {!!icon && (
        <>
          <AlertIcon {...iconStyleProps}>
            {icon}
          </AlertIcon>
          <Space minWidth="2x" />
        </>
      )}
      <AlertMessage {...messageStyleProps}>
        {children}
      </AlertMessage>
      {!!isClosable && (
        <>
          <Space minWidth="4x" />
          <AlertCloseButton {...closeButtonStyleProps} onClick={onClose}>
            <Icon icon="close-s" />
          </AlertCloseButton>
        </>
      )}
    </Flex>
  );
});

Alert.displayName = 'Alert';

export default Alert;
