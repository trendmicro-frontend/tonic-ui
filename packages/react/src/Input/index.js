import React, { forwardRef } from 'react';
import InputBase from '../InputBase';
import { useInputGroup } from '../InputGroup/context';
import { getInputGroupCSS, useInputStyle } from './styles';

const defaultSize = 'md';
const defaultVariant = 'outline';

const Input = forwardRef((
  {
    size,
    variant,
    css,
    ...rest
  },
  ref,
) => {
  const inputGroupContext = useInputGroup();
  if (inputGroupContext) {
    const {
      size: inputGroupSize,
      variant: inputGroupVariant,
    } = { ...inputGroupContext };

    // Use fallback values if values are null or undefined
    size = (size ?? inputGroupSize) ?? defaultSize;
    variant = (variant ?? inputGroupVariant) ?? defaultVariant;
    css = [
      getInputGroupCSS({ variant }),
      css,
    ];
  } else {
    // Use fallback values if values are null or undefined
    size = size ?? defaultSize;
    variant = variant ?? defaultVariant;
  }

  const styleProps = useInputStyle({ size, variant });

  return (
    <InputBase
      ref={ref}
      as="input"
      css={css}
      {...styleProps}
      {...rest}
    />
  );
});

Input.displayName = 'Input';

export default Input;
