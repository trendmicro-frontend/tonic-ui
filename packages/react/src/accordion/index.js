import Accordion from './Accordion';
import AccordionBody from './AccordionBody';
import AccordionCollapse from './AccordionCollapse';
import AccordionHeader from './AccordionHeader';
import AccordionItem from './AccordionItem';
import AccordionToggle from './AccordionToggle';
import AccordionToggleIcon from './AccordionToggleIcon';
import useAccordion from './useAccordion';
import useAccordionItem from './useAccordionItem';

Accordion.Body = AccordionBody;
Accordion.Collapse = AccordionCollapse;
Accordion.Header = AccordionHeader;
Accordion.Item = AccordionItem;
Accordion.Toggle = AccordionToggle;
Accordion.Toggle.Icon = AccordionToggleIcon;

export {
  Accordion,
  AccordionBody,
  AccordionCollapse,
  AccordionHeader,
  AccordionItem,
  AccordionToggle,
  AccordionToggleIcon,
  useAccordion,
  useAccordionItem,
};
