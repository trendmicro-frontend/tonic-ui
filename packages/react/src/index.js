// components
export * from './accordion';
export * from './alert';
export * from './badge';
export * from './box';
export * from './button';
export * from './checkbox';
export * from './color-mode';
export * from './color-style';
export * from './css-baseline';
export * from './divider';
export * from './drawer';
export * from './flat-button';
export * from './flex';
export * from './grid';
export * from './icon';
export * from './image';
export * from './input';
export * from './link';
export * from './menu';
export * from './modal';
export * from './pagination';
export * from './popover';
export * from './popper';
export * from './portal';
export * from './presence';
export * from './radio';
export * from './search-input';
export * from './scrollbar';
export * from './select';
export * from './skeleton';
export * from './space';
export * from './spinner';
export * from './stack';
export * from './switch';
export * from './table';
export * from './tabs';
export * from './tag';
export * from './text';
export * from './textarea';
export * from './theme';
export * from './toast';
export * from './tooltip';
export * from './transitions';
export * from './visually-hidden';

// hooks
export useDisclosure from './deprecated/useDisclosure'; // deprecated

// settings
export colorStyle from './shared/color-style';
export theme from './shared/theme';
