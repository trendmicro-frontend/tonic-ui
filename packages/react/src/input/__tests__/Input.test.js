import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render } from '@tonic-ui/react/test-utils/render';
import { testA11y } from '@tonic-ui/react/test-utils/accessibility';
import { Input } from '@tonic-ui/react/src';
import React from 'react';

describe('Input', () => {
  it('should render correctly', async () => {
    const sizes = ['sm', 'md', 'lg'];
    const variants = ['outline', 'filled', 'flush', 'unstyled'];
    const renderOptions = {
      useCSSVariables: true,
    };
    const { container } = render((
      <>
        {sizes.map(size => (
          variants.map(variant => (
            <Input
              data-testid={`input-${size}-${variant}`}
              key={`input-${size}-${variant}`}
              size={size}
              variant={variant}
            />
          ))
        ))}
      </>
    ), renderOptions);

    expect(container).toMatchSnapshot();

    await testA11y(container, {
      axeOptions: {
        rules: {
          label: { enabled: false },
        },
      },
    });
  });

  it('should render correctly with disabled attribute', () => {
    render(
      <Input data-testid="input" disabled />
    );
    const input = screen.getByTestId('input');
    expect(input).toHaveAttribute('aria-disabled', 'true');
  });

  it('should render correctly with error attributes', () => {
    render(
      <Input data-testid="input" error />
    );
    const input = screen.getByTestId('input');
    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  it('should render correctly with readonly attributes', () => {
    render(
      <Input data-testid="input" readOnly />
    );
    const input = screen.getByTestId('input');
    expect(input).toHaveAttribute('aria-readonly', 'true');
  });

  it('should render correctly with required attributes', () => {
    render(
      <Input data-testid="input" required />
    );
    const input = screen.getByTestId('input');
    expect(input).toHaveAttribute('aria-required', 'true');
  });

  it('should call the onChange function', async () => {
    const user = userEvent.setup();
    const onChange = jest.fn();
    render(
      <Input data-testid="input" onChange={onChange} />
    );
    const input = screen.getByTestId('input');
    await user.type(input, 'hello');
    expect(onChange).toHaveBeenCalled();
    expect(input).toHaveDisplayValue('hello');
  });

  it('should match the border color', async () => {
    const user = userEvent.setup();
    render(
      <Input data-testid="input" />
    );
    const input = screen.getByTestId('input');
    const defaultBorderColor = '#5e5e5e'; // [dark] gray:60 / [light] gray:30
    const focusBorderColor = '#1e5ede'; // [dark] blue:60 / [light] blue:60
    const hoverBorderColor = '#578aef'; // [dark] blue:50 / [light] blue:50

    expect(input).toBeValid();
    expect(document.body).toHaveFocus();

    await user.unhover(input);
    expect(input).toHaveStyleRule('border-color', defaultBorderColor);

    await user.hover(input);
    // XXX: Have to use `[data-hover]` to pass the assertion test
    expect(input).toHaveStyleRule('border-color', hoverBorderColor, { target: '[data-hover]' });

    await user.click(input);
    await user.unhover(input);
    expect(input).toHaveFocus();
    expect(input).toHaveStyleRule('border-color', focusBorderColor, { target: ':focus' });
  });
});
