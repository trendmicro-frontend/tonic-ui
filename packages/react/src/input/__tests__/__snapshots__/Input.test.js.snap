// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-6x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-0:invalid,
.emotion-0[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-0:invalid:focus:hover,
.emotion-0[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-0:invalid:hover,
.emotion-0[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-0::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-0::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-0:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-0::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-0:disabled,
.emotion-0:disabled:focus,
.emotion-0:disabled:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-gray-80);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-6x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-2:invalid,
.emotion-2[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-2:invalid:focus:hover,
.emotion-2[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-2:invalid:hover,
.emotion-2[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-2::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-2:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-2::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-2:focus,
.emotion-2[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-2:disabled,
.emotion-2:disabled:focus,
.emotion-2:disabled:hover,
.emotion-2[aria-disabled=true],
.emotion-2[aria-disabled=true]:focus,
.emotion-2[aria-disabled=true]:hover,
.emotion-2[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: none;
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-6x);
  border-bottom: var(--tonic-borders-1);
  border-bottom-color: var(--tonic-colors-gray-60);
}

.emotion-4:invalid,
.emotion-4[aria-invalid=true] {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-4:invalid:focus:hover,
.emotion-4[aria-invalid=true]:focus:hover {
  border-bottom-color: var(--tonic-colors-blue-60);
}

.emotion-4:invalid:hover,
.emotion-4[aria-invalid=true]:hover {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-4::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-4::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-4:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-4::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-4:focus,
.emotion-4[data-focus] {
  border-bottom-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-4:hover,
.emotion-4[data-hover] {
  border-bottom-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-4:disabled,
.emotion-4:disabled:focus,
.emotion-4:disabled:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover,
.emotion-4[data-disabled] {
  border-bottom-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-6 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: 0;
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
}

.emotion-8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-8x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-8:invalid,
.emotion-8[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-8:invalid:focus:hover,
.emotion-8[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-8:invalid:hover,
.emotion-8[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-8::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-8::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-8:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-8::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-8:focus,
.emotion-8[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-8:hover,
.emotion-8[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-8:disabled,
.emotion-8:disabled:focus,
.emotion-8:disabled:hover,
.emotion-8[aria-disabled=true],
.emotion-8[aria-disabled=true]:focus,
.emotion-8[aria-disabled=true]:hover,
.emotion-8[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-10 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-gray-80);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-8x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-10:invalid,
.emotion-10[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-10:invalid:focus:hover,
.emotion-10[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-10:invalid:hover,
.emotion-10[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-10::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-10::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-10:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-10::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-10:focus,
.emotion-10[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-10:hover,
.emotion-10[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-10:disabled,
.emotion-10:disabled:focus,
.emotion-10:disabled:hover,
.emotion-10[aria-disabled=true],
.emotion-10[aria-disabled=true]:focus,
.emotion-10[aria-disabled=true]:hover,
.emotion-10[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-12 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: none;
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-sm);
  line-height: var(--tonic-lineHeights-sm);
  height: var(--tonic-sizes-8x);
  border-bottom: var(--tonic-borders-1);
  border-bottom-color: var(--tonic-colors-gray-60);
}

.emotion-12:invalid,
.emotion-12[aria-invalid=true] {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-12:invalid:focus:hover,
.emotion-12[aria-invalid=true]:focus:hover {
  border-bottom-color: var(--tonic-colors-blue-60);
}

.emotion-12:invalid:hover,
.emotion-12[aria-invalid=true]:hover {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-12::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-12::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-12:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-12::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-12:focus,
.emotion-12[data-focus] {
  border-bottom-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-12:hover,
.emotion-12[data-hover] {
  border-bottom-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-12:disabled,
.emotion-12:disabled:focus,
.emotion-12:disabled:hover,
.emotion-12[aria-disabled=true],
.emotion-12[aria-disabled=true]:focus,
.emotion-12[aria-disabled=true]:hover,
.emotion-12[data-disabled] {
  border-bottom-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-16 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-md);
  line-height: var(--tonic-lineHeights-md);
  height: var(--tonic-sizes-10x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-16:invalid,
.emotion-16[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-16:invalid:focus:hover,
.emotion-16[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-16:invalid:hover,
.emotion-16[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-16::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-16::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-16:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-16::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-16:focus,
.emotion-16[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-16:hover,
.emotion-16[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-16:disabled,
.emotion-16:disabled:focus,
.emotion-16:disabled:hover,
.emotion-16[aria-disabled=true],
.emotion-16[aria-disabled=true]:focus,
.emotion-16[aria-disabled=true]:hover,
.emotion-16[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-18 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-gray-80);
  border: var(--tonic-borders-1);
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-md);
  line-height: var(--tonic-lineHeights-md);
  height: var(--tonic-sizes-10x);
  padding-left: calc(.75rem - .0625rem);
  padding-right: calc(.75rem - .0625rem);
  border-color: var(--tonic-colors-gray-60);
  border-radius: var(--tonic-radii-sm);
}

.emotion-18:invalid,
.emotion-18[aria-invalid=true] {
  border-color: var(--tonic-colors-red-50);
}

.emotion-18:invalid:focus:hover,
.emotion-18[aria-invalid=true]:focus:hover {
  border-color: var(--tonic-colors-blue-60);
}

.emotion-18:invalid:hover,
.emotion-18[aria-invalid=true]:hover {
  border-color: var(--tonic-colors-red-50);
}

.emotion-18::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-18::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-18:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-18::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-18:focus,
.emotion-18[data-focus] {
  border-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-18:hover,
.emotion-18[data-hover] {
  border-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-18:disabled,
.emotion-18:disabled:focus,
.emotion-18:disabled:hover,
.emotion-18[aria-disabled=true],
.emotion-18[aria-disabled=true]:focus,
.emotion-18[aria-disabled=true]:hover,
.emotion-18[data-disabled] {
  border-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-20 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: none;
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-md);
  line-height: var(--tonic-lineHeights-md);
  height: var(--tonic-sizes-10x);
  border-bottom: var(--tonic-borders-1);
  border-bottom-color: var(--tonic-colors-gray-60);
}

.emotion-20:invalid,
.emotion-20[aria-invalid=true] {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-20:invalid:focus:hover,
.emotion-20[aria-invalid=true]:focus:hover {
  border-bottom-color: var(--tonic-colors-blue-60);
}

.emotion-20:invalid:hover,
.emotion-20[aria-invalid=true]:hover {
  border-bottom-color: var(--tonic-colors-red-50);
}

.emotion-20::-webkit-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-20::-moz-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-20:-ms-input-placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-20::placeholder {
  color: var(--tonic-colors-white-tertiary);
  opacity: 1;
}

.emotion-20:focus,
.emotion-20[data-focus] {
  border-bottom-color: var(--tonic-colors-blue-60);
  z-index: 1;
}

.emotion-20:hover,
.emotion-20[data-hover] {
  border-bottom-color: var(--tonic-colors-blue-50);
  z-index: 2;
}

.emotion-20:disabled,
.emotion-20:disabled:focus,
.emotion-20:disabled:hover,
.emotion-20[aria-disabled=true],
.emotion-20[aria-disabled=true]:focus,
.emotion-20[aria-disabled=true]:hover,
.emotion-20[data-disabled] {
  border-bottom-color: var(--tonic-colors-gray-60);
  cursor: not-allowed;
  opacity: .28;
}

.emotion-22 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-transparent);
  border: 0;
  color: var(--tonic-colors-white-primary);
  outline: 0;
  padding: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: all .2s;
  transition: all .2s;
  width: 100%;
  font-size: var(--tonic-fontSizes-md);
  line-height: var(--tonic-lineHeights-md);
}

<div>
  <input
    class="emotion-0 emotion-1"
    data-testid="input-sm-outline"
  />
  <input
    class="emotion-2 emotion-1"
    data-testid="input-sm-filled"
  />
  <input
    class="emotion-4 emotion-1"
    data-testid="input-sm-flush"
  />
  <input
    class="emotion-6 emotion-1"
    data-testid="input-sm-unstyled"
  />
  <input
    class="emotion-8 emotion-1"
    data-testid="input-md-outline"
  />
  <input
    class="emotion-10 emotion-1"
    data-testid="input-md-filled"
  />
  <input
    class="emotion-12 emotion-1"
    data-testid="input-md-flush"
  />
  <input
    class="emotion-6 emotion-1"
    data-testid="input-md-unstyled"
  />
  <input
    class="emotion-16 emotion-1"
    data-testid="input-lg-outline"
  />
  <input
    class="emotion-18 emotion-1"
    data-testid="input-lg-filled"
  />
  <input
    class="emotion-20 emotion-1"
    data-testid="input-lg-flush"
  />
  <input
    class="emotion-22 emotion-1"
    data-testid="input-lg-unstyled"
  />
</div>
`;
